pkg/vmprovider/providers/vsphere2/virtualmachine/backup.go

```diff
@@ -28,136 +42,174 @@ type VMDiskData struct {
 			return err
 		}
 	}
 
 	return nil
 }
 
-func getDesiredVMKubeDataForBackup(
-	vm *vmopv1.VirtualMachine,
+// getDesiredResourceYAMLForBackup returns the encoded and gzipped YAML of the
+// given resources, or an empty string if the data is unchanged in ExtraConfig.
+func getDesiredResourceYAMLForBackup(
+	resources []client.Object,
+	ecKey string,
 	ecMap map[string]string) (string, error) {
-	// If the ExtraConfig already contains the latest VM spec, determined by
-	// 'metadata.generation', return an empty string to skip the backup.
-	if ecKubeData, ok := ecMap[vmopv1.VMBackupKubeDataExtraConfigKey]; ok {
-		vmFromBackup, err := constructVMObj(ecKubeData)
-		if err != nil {
-			return "", err
-		}
-		if vmFromBackup.ObjectMeta.Generation >= vm.ObjectMeta.Generation {
-			return "", nil
+	// Check if the given resources are up-to-date with the latest backup.
+	// This is done by comparing the resource versions of each object UID.
+	var isLatestBackup bool
+	if ecKubeData, ok := ecMap[ecKey]; ok {
+		if resourceToVersion := tryGetResourceVersion(ecKubeData); resourceToVersion != nil {
+			isLatestBackup = true
+			for _, curObj := range resources {
+				if curObj.GetResourceVersion() != resourceToVersion[string(curObj.GetUID())] {
+					isLatestBackup = false
+					break
+				}
+			}
 		}
 	}
 
-	backupVM := vm.DeepCopy()
-	backupVM.Status = vmopv1.VirtualMachineStatus{}
-	backupVMYaml, err := yaml.Marshal(backupVM)
-	if err != nil {
-		return "", err
+	// All resources are up-to-date, return an empty string to skip the backup.
+	if isLatestBackup {
+		return "", nil
+	}
+
+	// Backup the given resources YAML with encoding and compression.
+	// Use "---" as the separator if more than one resource is passed.
+	marshaledStrs := []string{}
+	for _, res := range resources {
+		marshaledYaml, err := k8syaml.Marshal(res)
+		if err != nil {
+			return "", fmt.Errorf("failed to marshal object %q: %v", res.GetName(), err)
+		}
+		marshaledStrs = append(marshaledStrs, string(marshaledYaml))
 	}
 
-	return util.EncodeGzipBase64(string(backupVMYaml))
+	resourcesYAML := strings.Join(marshaledStrs, "\n---\n")
+	return util.EncodeGzipBase64(resourcesYAML)
 }
 
-func constructVMObj(ecKubeData string) (vmopv1.VirtualMachine, error) {
-	var vmObj vmopv1.VirtualMachine
-	decodedKubeData, err := util.TryToDecodeBase64Gzip([]byte(ecKubeData))
+// tryGetResourceVersion tries to get the resource version of each object in
+// the encoded and gzipped YAML. Returns a map of resource UID to version.
+func tryGetResourceVersion(ecResourcesYAML string) map[string]string {
+	decoded, err := util.TryToDecodeBase64Gzip([]byte(ecResourcesYAML))
 	if err != nil {
-		return vmObj, err
+		return nil
 	}
 
-	err = yaml.Unmarshal([]byte(decodedKubeData), &vmObj)
-	return vmObj, err
+	resourceVersions := map[string]string{}
+	decUnstructured := yaml.NewDecodingSerializer(unstructured.UnstructuredJSONScheme)
+	resourcesYAML := strings.Split(decoded, "---\n")
```

Comment:
Why is the Split separator different when encoding/decoding YAML?

Reply:
Good catch, updated to match.